modules/openy_features/openy_prgf/modules/openy_prgf_class_sessions/openy_prgf_class_sessions.module

```diff
@@ -0,0 +1,217 @@
+<?php
+
+/**
+ * @file OpenY Paragraph Class Sessions module file.
+ */
+
+use Drupal\openy_session_instance\Entity\SessionInstance;
+use Drupal\openy_session_instance\SessionInstanceManager;
+use Drupal\node\Entity\Node;
+use Drupal\Core\Link;
+use Drupal\Core\Url;
+
+/**
+ * Implements hook_theme().
+ */
+function openy_prgf_class_theme() {
+  $items = [
+    'class_sessions' => [
+      'variables' => [
+        'session_instances' => [],
+      ],
+      'template' => 'openy-class-sessions',
+    ],
+  ];
+  return $items;
+}
+
+/**
+ * Implements hook_preprocess_paragraph__bundle().
+ */
+function openy_prgf_class_sessions_preprocess_paragraph__class_sessions(&$variables) {
```

Comment:
This is the abstraction layer that we really need for future customization and scalability.